libs/@local/hash-graph-types/graph_types/data_type.py

```diff
@@ -96,57 +95,27 @@ async def create_data_type(
         *,
         actor_id: UUID,
         graph: "GraphAPIProtocol",
-    ) -> Annotated[Any, "DataTypeAnnotation"]:
+    ) -> type[RootModel]:
         """Create an annotated type from this schema."""
+        # make sure that we don't duplicate work and create the same model twice
+        if self._cached is not None:
+            return self._cached
+
         # Custom data types will require an actor ID and the graph to be passed in
         _actor_id = actor_id
         _graph = graph
 
         const = self.model_extra.get("const") if self.model_extra else None
 
-        # TODO: Use `Field` instead when multiple fields are supported
-        #   https://github.com/pydantic/pydantic/issues/6349
-        #   https://github.com/pydantic/pydantic/issues/6353
-        class DataTypeAnnotation:
-            @classmethod
-            def __get_pydantic_core_schema__(
-                cls,
-                source_type: Any,  # noqa: ANN401
-                handler: GetCoreSchemaHandler,
-            ) -> CoreSchema:
-                schema = handler(source_type)
-                if const is not None:
-                    return core_schema.no_info_after_validator_function(
-                        cls.validate_const,
-                        schema,
-                    )
-                return schema
-
-            @classmethod
-            def __get_pydantic_json_schema__(
-                cls,
-                schema: CoreSchema,
-                handler: GetJsonSchemaHandler,
-            ) -> JsonSchemaValue:
-                json_schema = handler(schema)
-                json_schema.update(
-                    **{
-                        "$id": self.identifier,
-                        "$schema": self.schema_url,
-                        "title": self.title,
-                        "description": self.description,
-                        "kind": self.kind,
-                    },
-                )
-                if const is not None:
-                    json_schema.update(const=const)
-
-                return json_schema
-
-            @classmethod
-            def validate_const(cls, v: DataType) -> None:
-                if v != const:
-                    msg = f"Value must be {const}"
-                    raise ValueError(msg)
-
-        return Annotated[self._type(), DataTypeAnnotation]
+        type_ = Literal[const] if const is not None else self._type()
```

Comment:
`const` is not literal. This is also why Pylance is complaining on this line:
<img width="1455" alt="image" src="https://github.com/hashintel/hash/assets/21277928/bf1f734d-9e93-44f5-b636-4e000bc9350d">
In fact, I tried using `Literal` before but this is not allowed as the value is not a literal but could literally (🙃) be anything.
See https://github.com/pydantic/pydantic/issues/6346

Reply:
I tend to disagree. See this snippet:

```
from typing import Literal

from pydantic import TypeAdapter

a = "abc"

x = Literal[a]
print(x)

adapter = TypeAdapter(x)
adapter.validate_python("abcd")
```

```
Traceback (most recent call last):
  File "/Users/bmahmoud/Library/Application Support/JetBrains/PyCharm2023.1/scratches/scratch_41.py", line 11, in <module>
    adapter.validate_python("abcd")
  File "/Users/bmahmoud/Sync/projects/scionics/hash/client/.venv/lib/python3.11/site-packages/pydantic/type_adapter.py", line 206, in validate_python
    return self.validator.validate_python(__object, strict=strict, from_attributes=from_attributes, context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for literal['abc']
  Input should be 'abc' [type=literal_error, input_value='abcd', input_type=str]
    For further information visit https://errors.pydantic.dev/2.0.3/v/literal_error
```

Reply:
but it will fail if the value is a non-primitive during construction. Which is a bug anyway

Reply:
fixed in 9d30d651ed2a11e9333d3d57913227bf72d764e8